src/basic/conf-files.c

```diff
@@ -56,8 +57,35 @@ static int files_add(Hashmap *h, const char *root, const char *path, const char
         FOREACH_DIRENT(de, dir, return -errno) {
                 char *p;
 
-                if (!dirent_is_file_with_suffix(de, suffix))
+                if (!dirent_is_file_with_suffix(de, suffix)) {
+                        log_debug("Ignoring %s/%s, because it's not a regular file with suffix %s.", dirpath, de->d_name, strna(suffix));
                         continue;
+                }
+
+                if (flags & CONF_FILES_EXECUTABLE) {
+                        struct stat st;
+
+                        /* As requested: check if the file is marked exectuable. Note that we don't check access(X_OK)
+                         * here, as we care about whether the file is marked executable at all, and not whether it is
+                         * executable for us, because if such errors are stuff we should log about. */
+
+                        if (fstatat(dirfd(dir), de->d_name, &st, 0) < 0) {
+                                log_debug_errno(errno, "Failed to stat %s/%s, ignoring: %m", dirpath, de->d_name);
+                                continue;
+                        }
+
+                        /* We only want executable regular files (or symlinks to them), or symlinks to /dev/null */
```

Comment:
The patch looks good to me on the whole. The only thing I don't understand is why `files_add` passes names of empty executable files and symlinks to `/dev/null` to the callers. Would it be better to rearrange the `if` clause a bit, so that it always skips `null_and_empty`? I mean something like
```c
if (null_or_empty(...)) {
    ...
} else if (S_ISREG(st.st_mode)) {
   ...
} else {
   ...
}
```

Reply:
`null_or_empty()==true` is the exception, normally we expect to go down the first branch, i.e. to have a regular file. And we want to pass files that actually are `null_or_empty` to the callers because they mask other files. So unless I'm missing something, the check is good as is.

Reply:
`files_add` doesn't pass empty non-executable files to the callers and that seems a bit confusing to me.

Reply:
we need to collect the /dev/null symlinks to make the overriding logic work, even though we ultimately don't care about them: a symlink to /dev/null in /etc overrides a file of the same name in /usr, and hence we really can't skip it early on.

Reply:
Why can empty non-executables files be skipped early on?

Reply:
I've just filed https://github.com/systemd/systemd/issues/6831.

Reply:
Ouch, you are right. Thanks for filing the issue! (BTW, did I already mention how much I appreciate your sharp eye? Very much!)